### PR TITLE
OpenGL 2.0 backend

### PR DESCRIPTION
**Pull request by Beyley**

### CODE REVIEW

File: Furball.Vixie/Graphics/Backends/OpenGL20/OpenGL20Backend.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            if (this._maxTexUnits == -1)
+                this._maxTexUnits = this.gl.GetInteger((GLEnum)GetPName.MaxTextureImageUnits);
+
+            return this._maxTexUnits;
+        }
+        public override void Clear() {
+            this.gl.Clear(ClearBufferMask.ColorBufferBit);
+            this.gl.ClearColor(0.5f, 0, 0, 0);
+        }
+        public override TextureRenderTarget CreateRenderTarget(uint width,     uint height)      => new TextureRenderTargetGL20(this, width, height);
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/OpenGL20Backend.cs]: this autoformatting looks horrific

**Review by Eeveelution — CHANGES_REQUESTED** [review]

File: Furball.Vixie/Graphics/Backends/OpenGL20/OpenGL20Backend.cs (hunk 1 above)
Comment (Beyley) [reply Furball.Vixie/Graphics/Backends/OpenGL20/OpenGL20Backend.cs]: i think it looks fine but sure lemme change it

**Comment by Eeveelution** [comment]
i assume this is ready to merge?

**Review by Eeveelution — APPROVED** [review]

**Comment by Beyley** [comment]
Ive found a bug with our RenderTarget implementation in all GL backends, tracking it down rn

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                this.UnlockingUnbind();
+
+            if (this._isDisposed)
+                return;
+
+            this._isDisposed = true;
+
+            try {
+                this.ext.DeleteFramebuffer(this._frameBufferId);
+                this.gl.DeleteTexture(this._textureId);
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs]: Why is is DeleteTexture here and DeleteTextures in TextureGL20? im pretty sure GL2.0 doesn't even have a singular delete, from looking at the automatically generated GLAD header for GL2.0

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureGL20.cs [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+        /// <summary>
+        /// Creates a Texture with a single White Pixel
+        /// </summary>
+        public unsafe TextureGL20(OpenGL20Backend backend) {
+            this._backend = backend;
+            // this._backend.CheckThread();
+            
+            this.gl = backend.GetOpenGL();
+
+            this.TextureId = this.gl.GenTexture();
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureGL20.cs:90]: Singular generation ain't a thing, glGenTextures would be necessary here

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs (hunk 2 above)
Comment (Eeveelution) [reply Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs]: on second thought, is it maybe something silk is overloading?

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureGL20.cs (hunk 3 above)
Comment (Eeveelution) [reply Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureGL20.cs:90]: unless silk.net is overloading, singular glGenTexture doesn't exist for 2.0

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        /// <exception cref="Exception">Throws Exception if the Target didn't create properly</exception>
+        public unsafe TextureRenderTargetGL20(OpenGL20Backend backend, uint width, uint height) {
+            this._backend = backend;
+            // this._backend.CheckThread();
+
+            this.gl  = backend.GetOpenGL();
+            this.ext = backend.GetOpenGLFramebufferEXT();
+
+            //Generate and bind a FrameBuffer
+            this._frameBufferId = this.ext.GenFramebuffer();
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs:70]: again, unless silk.net is overloading, the extension GL_EXT_framebuffer_object doesn't provide a singular glGenFramebuffer

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+            this.ext.BindFramebuffer(FramebufferTarget.Framebuffer, this._frameBufferId);
+            this._backend.CheckError();
+
+            //Generate a Texture
+            this._textureId = this.gl.GenTexture();
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs:75]: same for the texture

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            this.gl.BindTexture(TextureTarget.Texture2D, this._textureId);
+            //Set it to Empty
+            this.gl.TexImage2D(TextureTarget.Texture2D, 0, InternalFormat.Rgb, width, height, 0, PixelFormat.Rgb, PixelType.UnsignedByte, null);
+            //Set The Filtering to nearest (apperantly necessary, idk)
+            this.gl.TexParameterI(TextureTarget.Texture2D, GLEnum.TextureMagFilter, (int)GLEnum.Nearest);
+            this.gl.TexParameterI(TextureTarget.Texture2D, GLEnum.TextureMinFilter, (int)GLEnum.Nearest);
+            this._backend.CheckError();
+
+            //Generate the Depth buffer
+            this._depthRenderBufferId = this.ext.GenRenderbuffer();
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs:85]: same for this renderbuffer

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            if (this.Bound)
+                this.UnlockingUnbind();
+
+            if (this._isDisposed)
+                return;
+
+            this._isDisposed = true;
+
+            try {
+                this.ext.DeleteFramebuffer(this._frameBufferId);
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs:219]: same with the singular thing

File: Furball.Vixie/Graphics/Backends/OpenGL20/LineRendererGL20.cs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        internal unsafe LineRendererGL20(OpenGL20Backend backend) {
+            this._backend = backend;
+            this._gl      = backend.GetOpenGL();
+            
+            string vertex   = ResourceHelpers.GetStringResource("ShaderCode/OpenGL20/LineVertexShader.glsl");
+            string fragment = ResourceHelpers.GetStringResource("ShaderCode/OpenGL20/LineFragmentShader.glsl");
+
+            this._program = new ProgramGL20(backend, vertex, fragment);
+
+            this._arrayBuf = this._gl.GenBuffer();
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/LineRendererGL20.cs:34]: singular thing

File: Furball.Vixie/Graphics/Backends/OpenGL20/QuadRendererGL20.cs [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+                BatchedIndicies[i * 6 + 2] = (ushort)(2 + i * 4);
+                //Top right
+                BatchedIndicies[i * 6 + 3] = (ushort)(2 + i * 4);
+                //Bottom left
+                BatchedIndicies[i * 6 + 4] = (ushort)(3 + i * 4);
+                //Bottom right
+                BatchedIndicies[i * 6 + 5] = (ushort)(0 + i * 4);
+            }
+
+            this._vertexBuffer = this._gl.GenBuffer();
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/QuadRendererGL20.cs:111]: singular thing

**Review by Eeveelution — CHANGES_REQUESTED** [review]
It pretty much boils down to that OpenGL 2.0 didn't have singular variants for the glGen* functions, glCreateShader and glCreateProgram do exist as singular, but glGenBuffer and glGenFramebuffer and stuff, even in the GL_EXT_framebuffer_object extension don't exist. They only exist in their plural form like glGenBuffers or glGenFramebuffers, which generate multiple

My only hypothesis as to why this may be working in software on emulated OpenGL 2.0 is that Silk.NET might be overloading this to just create a single buffer while calling the multiple form internally, if you can check up on that and either make sure Silk.NET is doing that, or changing the code to fit what I'm seeing, I'll approve.

**Comment by Eeveelution** [comment]
If you're curious as to what API calls exist in 2.0 with the GL_EXT_framebuffer_object extension, I uploaded the automatically generated GLAD header here: https://eevee-sylveon.s-ul.eu/cAbZOvEo
You can generate those yourself here: https://glad.dav1d.de/ they generate headers and loaders for various OpenGL versions according to the OpenGL specs, while they may be for C/C++, it's still good to  have some sort of overview of what exists and what doesn't when looking into the glad.h header file located in `include/glad/glad.h`

File: Furball.Vixie/Graphics/Backends/OpenGL20/QuadRendererGL20.cs [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+
+        private readonly ProgramGL20 _program;
+        private readonly uint        _vertexBuffer;
+
+        private bool disposed = true;
+        public void Dispose() {
+            if (this.disposed) return;
+            
+            this._program.Dispose();
+            this._gl.DeleteBuffer(this._vertexBuffer);
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/QuadRendererGL20.cs:24]: Same with here with the singular

File: Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs (hunk 2 above)
Comment (Eeveelution) [reply Furball.Vixie/Graphics/Backends/OpenGL20/Abstractions/TextureRenderTargetGL20.cs]: Conslusion: Silk.NET does overloads! Although i feel like at least TextureGL20 should be consistent with everywhere else, Texture20GL does glDeleteTextures while everywhere else its glDeleteTexture

**Comment by Eeveelution** [comment]
Conslusion: Silk.NET does overloads!

**Review by Eeveelution — APPROVED** [review]
Nice, that makes OpenGL 2.0 the 3rd officially supported Backend in Vixie!

**Pull request merged**